Added advanced Heavy Hitter Detection example

gnikol opened the pull request:
Heavy hitter tutorial used for CS344 in Stanford.

jnfoster commented:
This is cool. 

We don't really keep code in the top-level directory. And it'd be nice to harmonize it with the build processes used for other tutorials. 

Would it be possible to make these changes before we merge?

gnikol commented:
Sure. I'm not sure about where the tutorial should go that's why I put on the top directory waiting for suggestions. It obviously cannot go under the event-related directories so, should I put it under examples?

Regarding the building, my code uses p4app, which I think is also used by some of the other tutorials. Could you be more specific about what I should do?

gnikol commented:
- I added the same same build process like in SIGCOMM 2017 (e.g. HULA).
- Fixed the skeleton program
- Added comments in the TODO sections
- Updated README

gnikol commented on Advanced_HHD/README.md:
```diff
@@ -0,0 +1,31 @@
+# Instructions
+
+You have to fill in the parts of simple_router.p4 marked with "TODO". You can find a solution in the solution folder.
```
Done

antoninbas commented on Teaching/Stanford_CS344_2018/README.md:
```diff
@@ -0,0 +1,126 @@
+matches a packet, its action is invoked with parameters supplied by
+the control plane as part of the rule.
+
+In this exercise, we have already implemented the the control plane
```
typo

gnikol replied:
Fixed

antoninbas commented on Teaching/Stanford_CS344_2018/README.md:
```diff
@@ -0,0 +1,126 @@
+file, you have to fill in the parts that are marked with TODO.
+
+This example is based on [count-min sketch](http://theory.stanford.edu/~tim/s15/l/l2.pdf).
+In fact, we use two count-mi sketches which are reset with an offset
```
typo

gnikol replied:
Fixed

antoninbas commented on Teaching/Stanford_CS344_2018/README.md:
```diff
@@ -0,0 +1,126 @@
+
+The `simple_router.p4` file contains a skeleton P4 program with key pieces of
+logic replaced by `TODO` comments. Your implementation should follow
+the structure given in this file---replace each `TODO` with logic
```
does the `---` mean something?

gnikol replied:
Fixed

antoninbas commented on Teaching/utils/mininet/appcontroller.py:
```diff
@@ -0,0 +1,104 @@
+                    entries[sw] += extra_entries
+                else: # path to file that contains entries
+                    entries[sw] += self.read_entries(extra_entries)
+            #entries[sw] += [
```
I would recommend just removing what you don't need (it seems that these are read from the file)

gnikol replied:
Fixed

antoninbas commented on Teaching/Stanford_CS344_2018/simple_router.config.template:
```diff
@@ -0,0 +1,12 @@
+set_crc16_parameters calc_2 0x1021 0xffff 0x0000 false false
```
I'm not too familiar with p4app, is there a reason for the `.template` suffix in the file name?

gnikol replied:
p4app needs simple_router.config which generated by setup.py by appending a set of rules to the contents of simple_router.config.template.

gnikol commented:
@jnfoster any more input, or we can proceed with merging?
